DistributedTransaction/DistributedTransaction.tla

```diff
@@ -269,42 +385,117 @@ ServerLockKey ==
         start_ts == req.start_ts
        IN
         \* Pessimistic lock is allowed only if no stale lock exists.  If
-        \* there is one, wait until ServerCleanupStaleLock to clean it up.
-        /\ key_lock[k] = {}
-        /\ LET
-              latest_write == {w \in key_write[k] : \A w2 \in key_write[k] : w.ts >= w2.ts}
-              
-              all_commits == {w \in key_write[k] : w.type = "write"}
-              latest_commit == {w \in all_commits : \A w2 \in all_commits : w.ts >= w2.ts}
-           IN
-              IF \E w \in key_write[k] : w.start_ts = start_ts /\ w.type = "rollback"
+        \* there is one, wait until ClientCheckTxnStatus to clean it up.
+        IF key_lock[k] = {} THEN
+          /\ LET
+                latest_write == {w \in key_write[k] : \A w2 \in key_write[k] : w.ts >= w2.ts}
+                
+                all_commits == {w \in key_write[k] : w.type = "write"}
+                latest_commit == {w \in all_commits : \A w2 \in all_commits : w.ts >= w2.ts}
+                commit_read == {w \in all_commits : \A w2 \in all_commits : w2.ts <= w.ts \/ w2.ts >= start_ts}
+             IN
+                IF \E w \in key_write[k] : w.start_ts = start_ts /\ w.type = "rollback"
+                THEN
+                  \* If corresponding rollback record is found, which
+                  \* indicates that the transcation is rollbacked, abort the
+                  \* transaction.
+                  /\ SendResp([start_ts |-> start_ts, type |-> "lock_key_aborted"])
+                  /\ UNCHANGED <<req_msgs, client_vars, key_vars, next_ts>>
+                ELSE
+                  \* Acquire pessimistic lock only if for_update_ts of req
+                  \* is greater or equal to the latest "write" record.
+                  \* Because if the latest record is "write", it means that
+                  \* a new version is committed after for_update_ts, which
+                  \* violates Read Committed guarantee.
+                  \/ /\ ~ \E w \in latest_commit : w.ts > req.for_update_ts
+                     /\ key_lock' = [key_lock EXCEPT ![k] = {[ts |-> start_ts,
+                                                              primary |-> req.primary,
+                                                              min_commit_ts |-> NoneTs,
+                                                              type |-> "lock_key"]}]
+                     /\ IF ~ commit_read = {}
+                        THEN
+                          LET
+                           commit_record == CHOOSE value \in commit_read : TRUE
+                           value_ts == commit_record.start_ts
+                          IN
+                          /\ \/ /\ ClientLockedKey([start_ts |-> start_ts, 
+                                                    type |-> "locked_key", 
+                                                    key |-> k, 
+                                                    value_ts |-> value_ts])

```

Comment:
Where is value_ts used?

Reply:
there's no direct use since we don't care about the actual value, but the isolation level of pessimistic read needs it.